### DISCUSSION TITLE
Need directions and advice for mixing content with Razor pages.

### DISCUSSION THREAD

**Discussion by DrHeinzDoofenshmirtz** [OrchardCMS/OrchardCore#9846] [Q&A]
Hi everyone.

I have been exploring CMS solutions for .Net Core and Orchard Core seems like one of the best free options.
Though, I have some questions, as I find the data model of Orchard to be quite confusing, and I can't seem to figure out, if I can do what I want in Orchard in an easy way. I imagine the Decoupled architecture makes the most sense.

Basically, the minimum I want is just an external place where I can edit the text in my Razor pages. So that every time I have to edit a Razor page's text, I don't have to do a new deploy. Furthermore, I would like some good Localization functionality on top of that. And as I also myself create some articles to explain the concepts of my website, that is also a good reason to use a CMS.
That is really what I need. The rest I can do in Razor pages.

The site also has some custom functionality, and I will use the idea of a wishlist site to explain.
Users can create wishlists that will be stored in a database with a reference to their user id.
When the user accesses the route "/wishlists" a table or list is shown with the names of they wishlists, with options to edit them, create a new one, etc.
The table is easy enough to create in a Razor page, as it is simply a loop over the user's wishlists.
If I also had some text on this page, let's say:

...
TEXT
WISHLIST TABLE
TEXT
...

How would I go ahead making this in Orchard Core?
Is the wishlist table a Widget? And if so, how do I create a Widget that gets its code from a Razor page?
How can I mix content from an Orchard Core page with content from a Razor page?

I hope you can put me on the right track. Thanks.

Best
Matthias

**Comment by ns8482e** [comment]
There are different ways, you can store your data in OC. You can create tenant aware tables and use apis provided in `OrchardCore.Data` module to query your tables. You can also manage your data as `Content` by creating  `ContentType` using `ContentFields` and `ContentParts`

OC is multi-tenant - modular framework so it's best to organize your application in modules and make it tenant aware. To use it CMS as backend in decoupled site start with following .

- Create a module - [how to](https://docs.orchardcore.net/en/dev/docs/getting-started/templates/index.html#create-a-new-cms-module)
- Create a decoupled site [how-to](https://docs.orchardcore.net/en/dev/docs/guides/decoupled-cms/index.html)
- Create a content type - [how-to](https://docs.orchardcore.net/en/dev/docs/guides/decoupled-cms/index.html#creating-a-blog-post-content-type)
- Use Migration to create content type dynamically without using UI - [how-to](https://docs.orchardcore.net/en/dev/docs/reference/modules/ContentTypes/index.html#migrations)
- Query your data - [how-to](https://docs.orchardcore.net/en/dev/docs/reference/core/Data/index.html#running-sql-queries)
- Use Orchard [Razor Helper ](https://docs.orchardcore.net/en/dev/docs/reference/core/Razor/index.html#razor-helpers)-

**Reply by DrHeinzDoofenshmirtz** [reply]
Hi.
Thanks for the reply.
I am not really sure if I understand, what you are explaining.

If I make it tenant aware and use `OrchardCore.Data`, is that for a decoupled site?
And managing the data as `Content`, is that what I am talking about when making a Widget?

**Reply by ns8482e** [reply]
> I am not really sure if I understand, what you are explaining.

It really depends how you want to manage your data in OC

In Orchard Core CMS - `Contents` are `ContentItem` and these content items are stored in tenant aware tables i.e. Tables in SQL are named with tenant specific table prefix. So If you have Tenant prefix `TC` then table will be `TC_Document`. You can work with your content item using `IOrchardHelper` and  `IContentManager` and you can also query content using GraphQL/Lucene.

If you want to manage your data (i.e.  WishList) as `ContentType`  and want to utilize other content management features then you can create `ContentType`  and use OC CMS' admin area to create/edit/manage your `ContentItem` and use `IOrchardHelper` Razor helper to get the data in your decoupled site pages.

If you do not want to store your data as `ContentType` and want to manage your data in your own SQL tables (make it tenant aware by adding table prefix for the tenant) and want to perform CRUD operations using your own Pages/Controllers then you can query your data as explained in `OrchardCore.Data` [documentation](https://docs.orchardcore.net/en/dev/docs/reference/core/Data/index.html).

**Reply by DrHeinzDoofenshmirtz** [reply]
@ns8482e Okay, that makes sense.
I was planning on using an external database, which I think is the best solution for me. I wasn't planning on having OC manage everything else than text and image content.

But my questions still stands. How do I combine that? How do I make "views" with text and images stored in OC that can also show a table that is created by a Razor Page?

Thanks for your help so far!

**Reply by ns8482e** [reply]
@DrHeinzDoofenshmirtz in decoupled development approach back-end is driven by a CMS and rendering is controlled by pages/view on front-end.

To use orchard core in decoupled way you can inject  `IOrchardHelper` to get the data as explained [in guide](https://docs.orchardcore.net/en/dev/docs/guides/decoupled-cms/index.html).

For example, following code will get content by id in razor page and log data in console
```cshtml
@page "/blogpost/{id}"
@inject OrchardCore.IOrchardHelper Orchard

@{
    var blogPost = await Orchard.GetContentItemByIdAsync(Id);
}

@Orchard.ConsoleLog(blogPost)
```
Once you have `blogPost` object you can render it however you want it to be displayed in razor page.
